test-consensus/Test/DynamicBFT.hs

```diff
@@ -277,24 +280,18 @@ broadcastNetwork numSlots nodeInit txMap initLedger initRNG = do
       varSt  <- atomically $ newTVar initSt
       varL   <- atomically $ newTVar initLedger
 
-      let ourOwnConsumer :: Chan m (MsgConsumer (SimpleBlock p c))
-                                   (MsgProducer (SimpleBlock p c))
+      let ourOwnConsumer :: Channel m (SomeTransition (ChainSyncMessage (SimpleBlock p c) (Point (SimpleBlock p c))))
           ourOwnConsumer =
               let loopbackRecv = snd (chans Map.! us Map.! us)
-              in Chan {
-                     sendMsg = sendMsg loopbackRecv
-                   , recvMsg = do
-                       msgToMyself <- recvMsg loopbackRecv
-                       case msgToMyself of
-                           MsgRollForward b -> do
-                               atomically $
-                                 modifyTVar' varL $ \st ->
-                                   case runExcept (applyExtLedgerState cfg b st) of
-                                     Left err  -> error (show err)
-                                     Right st' -> st'
-                           _ -> return ()
-                       return msgToMyself
-                  }
+                  recvEff mMsgToMyself = case mMsgToMyself of
+                    Just (SomeTransition (MsgRollForward b _)) -> do
+                        atomically $
+                          modifyTVar' varL $ \st ->
+                            case runExcept (applyExtLedgerState cfg b st) of
+                              Left err  -> error (show err)
+                              Right st' -> st'
+                    _ -> return ()
+              in channelRecvEffect recvEff loopbackRecv
```

Comment:
This block, both before and after the change, looks very very dubious to me. As far as I understand it, this makes the _channel_ from this node to itself do some update to the ledger. That's just completely contrary to what we want: the protocol application should do this. In fact, the node should not even have a loopback. Why produce and consume from one's self? I see that below on line 315 there's a comment

> TODO: We *do not* update the ledger state here. That's done in
> our \"loopback\" consumer.

Really it's done in the loopback _channel_, which is not a consumer. It's supposed to carry bytes or messages and that's all.

What are we missing here in order to do it the right way? We want a protocol application which does the `applyExtLedgerState`, and which is _driven_ by this channel, which can simply push a `SomeTransition (ChainSyncMessage (SimpleBlock p c) (Point (SimpleBlock p c)))` into the application. We have `relayNode` which gives a `TVar m (ChainProducerState block)`. Can we not use that to do `applyExtLedgerState`? Whenever the chain changes, we un-apply all of the blocks in the old fork not in the new fork (if any), and apply all of the blocks in the new fork now in the old.

Is what we have now even right? Shouldn't it also un-apply on a `MsgRollBackward`?

@edsko @adinapoli-iohk

Reply:
This is completely changed in my branch; I'm just waiting for this PR to land and then I'll rebase that work. I'll ping you on that PR to see if you still feel we're abusing the network layer or not :)